src/clients/c++/experimental_api_v2/library/common.h

```diff
@@ -0,0 +1,561 @@
+  /// will handle the request using default setting for the model.
+  uint64_t priority_;
+  /// The timeout value for the request, in microseconds. If the request
+  /// cannot be completed within the time the server can take a
+  /// model-specific action such as terminating the request. If not
+  /// provided, the server will handle the request using default setting
+  /// for the model.
+  uint64_t timeout_;
+};
+
+//==============================================================================
+/// An interface for InferInput object to describe the model input for
+/// inference.
+///
+class InferInput {
+ public:
+  /// Create a InferInput instance that describes a model input.
+  /// \param infer_input Returns a new InferInput object.
+  /// \param name The name of input whose data will be described by this object.
+  /// \param dims The shape of the input.
+  /// \param datatype The datatype of the input.
+  /// \return Error object indicating success or failure.
+  static Error Create(
+      InferInput** infer_input, const std::string& name,
+      const std::vector<int64_t>& dims, const std::string& datatype);
+
+  /// Gets the name of input being described by this object
+  /// \return The name of the input
+  const std::string& Name() const { return name_; }
+
+  /// Gets the size of data set into this input in bytes.
+  /// \return The size of data in bytes.
+  int64_t ByteSize() const { return byte_size_; }
+
+  /// Gets the size of data which will be send on the wire.
+  /// \return The size of data to be sent on wire in bytes.
+  size_t TotalSendByteSize() const { return total_send_byte_size_; }
+
+  /// Gets the dataype of the input.
+  /// \return The datatype of this input.
+  const std::string& Datatype() const { return datatype_; }
+
+  /// Gets the current shape of the input.
+  /// \return The shape of this input.
+  const std::vector<int64_t>& Shape() const { return shape_; }
+
+  /// Returns whether the input is set with shared memory or not.
+  /// \return The truth value of wether input is set with shared
+  /// memory.
+  bool IsSharedMemory() const { return (io_type_ == SHARED_MEMORY); }
+
+  /// Set the shape of input associated with this object.
+  /// \param dims the vector of dims representing the new shape
+  /// of input.
+  /// \return Error object indicating success or failure of the
+  /// request.
+  Error SetShape(const std::vector<int64_t>& dims);
+
+  /// Prepare this input to receive new tensor values. Forget any
+  /// existing values that were set by previous calls to SetSharedMemory()
+  /// or SetRaw().
+  /// \return Error object indicating success or failure.
+  Error Reset();
+
+  /// Set tensor values for this input from a byte vector. The vector
+  /// is not copied and so it must not be modified or destroyed
+  /// until this input is no longer needed (that is until the Run()
+  /// call(s) that use the input have completed). For batched inputs
```

Comment:
For batched inputs it would be nice to not have to say that it must be called once for each batch... really we just want them to be able to provide input in any types of chunks they want. It there anything in the implementation that relies on it being called exactly batch-size times?